Add blog: Detailed explanation of SkyWalking Go Toolkit Trace (#650)
apache/skywalking-website: content/blog/2023-10-18-skywalking-toolkit-trace/index.md

Before diving into this, you can learn how to use the Skywalking Go agent by referring to the [SkyWalking Go Agent Quick Start Guide](https://skywalking.apache.org/zh/2023-06-01-quick-start-with-skywalking-go-agent/).

The following sections will explain how to use these interfaces in specific scenarios.

## Introducing the Trace Toolkit

Execute the following command in the project's root directory:

```shell
go get github.com/apache/skywalking-go/toolkit
```

To use the toolkit trace interface, you need to import the package into your project:

```go
"github.com/apache/skywalking-go/toolkit/trace"
```

## Manual Tracing

A Span is the fundamental unit of an operation in Tracing. It represents an operation within a specific timeframe, such as a request, a function call, or a specific action. It records essential information about a particular operation, including start and end times, the operation's name, tags (key-value pairs), and relationships between operations. Multiple Spans can form a hierarchical structure.

In situations where Skywalking-go doesn't support a particular framework, users can manually create Spans to obtain tracing information.

> (Here, I have removed the originally supported frameworks for the sake of the example. If needed, please replace them with unsupported frameworks.)

Reviewer: ```suggestion
```suggestion
(Here, I have removed the supported frameworks for the sake of the example. These are only examples. You should reference this when using the APIs in private and/or unsupported frameworks)
```
```

		return r.Header.Get(headerKey), nil
	})
	// Finish the EntrySpan.
	defer trace.StopSpan()

	_, _ = w.Write([]byte("success from provider"))
}

func main() {
	http.HandleFunc("/provider", providerHandler)

	_ = http.ListenAndServe(":9998", nil)
}

```

Then, in the terminal, execute:

```shell
go build -toolexec="/path/to/go-agent" -a -o consumer ./consumer.go
./consumer
```

```shell
go build -toolexec="/path/to/go-agent" -a -o provider ./provider.go
./provider
```

```shell
curl 127.0.0.1:9999/consumer
```

At this point, the UI will display the span information you created.

![Trace List](./trace.png)

If you need to trace methods that are executed only locally, you can use trace.CreateLocalSpan(). If you don't need to monitor information or states from the other end, you can change ExitSpan and EntrySpan to LocalSpan.

The usage examples provided are for illustration purposes, and users can decide the tracing granularity and where in the program they need tracing.

Please note that if a program ends too quickly, it may cause tracing data to be unable to be asynchronously sent to the SkyWalking backend.

> ## Get/Set Tracing Data

Reviewer: ```suggestion
## Populate The Span
```

> Please note that if a program ends too quickly, it may cause tracing data to be unable to be asynchronously sent to the SkyWalking backend.

Reviewer 2: I am not sure if it is possible but can a manual sync be forced for short running programs?

Reviewer 3: Currently, skywalking-go does not support the feature of forced synchronization.

Reviewer: Sebastian asked a similar question on the slack channel too in PoC codes.

And I agree, we should keep async as always, and no plan to add that.

> When there's a necessity to record additional information, update, or modify the tags, logs, or operation names of the current traced Span, the Get/Set Tracing Data operations come into play. These actions are used to enhance trace information, providing a more detailed and precise contextual description, which aids in better understanding the events or operations being traced.

Reviewer: ```suggestion
When there's a necessity to record additional information, including creating/updating tags, appending logs, and setting a new operation name of the current traced Span, these APIs should be considered. These actions are used to enhance trace information, providing a more detailed and precise contextual description, which aids in better understanding the events or operations being traced.
```

> Toolkit trace provides a convenient way to access and manipulate trace data, including:

Reviewer: ```suggestion
Toolkit trace APIs provide a convenient way to access and manipulate trace data, including:
```

- Setting Tags: `SetTag()`
- Adding Logs: `AddLog()`
- Setting Span Names: `SetOperationName()`
- Getting various IDs: `GetTraceID()`, `GetSegmentID()`, `GetSpanID()`

For example, if you need to record the HTTP status code in a span, you can use the following interfaces while the span is not yet finished:

```go
trace.CreateExitSpan("GET:/provider", "localhost:9999", func(headerKey, headerValue string) error {
	r.Header.Add(headerKey, headerValue)
	return nil
})
resp, err := http.Get("http://localhost:9999/provider")

> trace.SetTag("status_code", fmt.Sprintf("%d", resp.StatusCode))

Reviewer 2: Stupid question from my side, how do we make sure that concurrency does not interfere with the current trace? Do we keep that in the context?

Reviewer 3: There is [a context propagation mechanism](https://skywalking.apache.org/docs/skywalking-go/next/en/concepts-and-designs/key-principles/#context-propagation-between-methods) to keep the trace. For short, We would obtain real-time trace data from the current goroutine when setting the tag.

spanID := trace.GetSpanID()
trace.StopSpan()
```

It's important to note that when making these method calls, the current thread should have an active span.

## Async APIs

> Async API for Span Manipulation in Another Goroutine.When there is a necessity to manipulate a traced Span in another goroutine, the use of the Async API becomes essential. These scenarios might include:

Reviewer: ```suggestion
Async APIs work for manipulating spans across Goroutines. These scenarios might include:
```

- Applications involving concurrency or multiple goroutines where operating on Spans across different execution contexts is necessary.
- Updating or logging information for a Span during asynchronous operations.
- Requiring a delayed completion of a Span.

To use it, follow these steps:

- Obtain the return value of CreateSpan, which is `SpanRef`.
- Call `spanRef.PrepareAsync()` to prepare for operations in another goroutine.

> - When the current goroutine's work is done, call `trace.StopSpan()` to end the span (affecting only in the current goroutine).

Reviewer 2: What happens if I pass the span ref to multiple go routines, does each one of them need to call `AsyncFinish()`?

Reviewer 3: No, you just need to call the `AsyncFinish` in one goroutine. If you call it multiple times, it would throw an error, right @Alipebt?

Author: Yes, it will throw an error about having closed it

- Pass the `spanRef` to another goroutine.
- After the work is done in any goroutine, call `spanRef.AsyncFinish()`.

Here's an example:

```go

> spanRef, _ := trace.CreateLocalSpan("LocalSpan")

Reviewer 2: Maybe do not ignore the second parameter of `CreateLocalSpan`. In case this is an error, have some default error handling for the example code.

Reviewer 3: @Alipebt Please add the error handler.

Correlation Context is commonly applied in the following scenarios:

- `Passing Information Between Spans`: It facilitates the transfer of critical information between different Spans, enabling upstream and downstream Spans to understand the correlation and context between each other.
- `Passing Business Parameters`: In business scenarios, it involves transmitting specific parameters or information between different Spans, such as authentication tokens, business transaction IDs, and more.

Users can set the Correlation Context using `trace.SetCorrelation(key, value)` and then retrieve the corresponding value in downstream spans using `value := trace.GetCorrelation(key)`.

For example, in the code below, we store the value in the tag of the span, making it easier to observe the result:

```go
package main

import (
	_ "github.com/apache/skywalking-go"
	"github.com/apache/skywalking-go/toolkit/trace"
	"net/http"
)

func providerHandler(w http.ResponseWriter, r *http.Request) {
	ctxValue := trace.GetCorrelation("key")
	trace.SetTag("result", ctxValue)
}

func consumerHandler(w http.ResponseWriter, r *http.Request) {

> 	trace.SetCorrelation("key", "value")

Reviewer 2: What kind of key and value shall be provided for the correlation? Do we have a best practice?

Reviewer 3: This depends on the needs of your business system. You just need to ensure that the content you want to transmit does not exceed the limits set in the configuration.

Reviewer: Actually, end users would choose. Sometimes, they put order ID, product ID, etc.
